### PR TITLE
[Tooling] add `go_lint` & `go_imports` make targets & CI step

### PR DESCRIPTION
**Pull request by bryanchriswhite**
### Human Summary

- Updates `install_ci_deps` make target to install `golangci-lint` and `goimports`
- Adds .golangci.yml config file; poised to progressively enable linters
- Adds  `go_lint` make target to run `golangci-lint`
- Adds  `go_imports` make target to run `goimports` via a go script: /tools/scripts/goimports.go
- Fixes non-excluded goimport issues
- Add stubs for mock packages that were causing CI to fail & cleanup .gitkeep files

## Issue

Related to:
- #128 

#128 surfaces some goimports issues and touches basically all imports already; might as well fix them all.

## Type of change

Select one or more:

- [ ] New feature, functionality or library
- [ ] Bug fix
- [x] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

- [ ] **Run all unit tests**: `make go_develop_and_test`
- [ ] **Verify Localnet manually**: See the instructions [here](TODO: add link to instructions)

## Sanity Checklist

- [x] I have tested my changes using the available tooling
- [ ] I have performed a self-review of my own code
- [x] I have commented my code, updated documentation and left TODOs throughout the codebase

### CODE REVIEW

File: .golangci.yml [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+#    - revive
+#    - errcheck
+#    - unused
+    - goimports
+
+issues:
+  exclude-use-default: true
+  max-issues-per-linter: 0
+  max-same-issues: 0
+  # TODO_CONSIDERATION/TODO_TECHDEBT: Perhaps we should prefer enforcing the best
```
Comment (Olshansk) [inline .golangci.yml:20]: I feel like at this point we need to send a PR to cosmos :)

File: internal/mocks/mockclient/mocks.go [hunk 2]
```diff
@@ -0,0 +1,5 @@
+package mockclient
```
Comment (Olshansk) [inline internal/mocks/mockclient/mocks.go:1]: Haha, I literally ended up having to do the same thing. Feel free to combine the comments after I merge mine in.

File: tools/scripts/goimports/main.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+	"bytes"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+)
+
+// defaultArgs are always passed to goimports.
+// -w: write result to (source) file instead of stdout
```
Comment (Olshansk) [inline tools/scripts/goimports/main.go:14]: OPTIONAL NIT: Replace `-w` with `-i` so its more similar to `sed`'s "in place" flag.

File: tools/scripts/goimports/main.go [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+		// Skip directories
+		if info.IsDir() {
+			return nil
+		}
+
+		// Add eligible Go files to the list
+		if filepath.Ext(path) == ".go" &&
+			// Ignore generated protobuf & protobuf gateway files.
+			!(strings.HasSuffix(path, ".pb.go") || strings.HasSuffix(path, ".pb.gw.go")) {
```
Comment (Olshansk) [inline tools/scripts/goimports/main.go:46]: Should we also skip `_mock.go`?

File: tools/scripts/goimports/main.go [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+// -local: put imports beginning with this string after 3rd-party packages (comma-separated list)
+// (see: goimports -h)
+var defaultArgs = []string{"-w", "-local", "github.com/pokt-network/poktroll"}
+
+func main() {
+	root := "."
+	var filesToProcess []string
+
+	// Walk the file system and accumulate matching files
+	err := filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
```
Comment (Olshansk) [inline tools/scripts/goimports/main.go:24]: OPTIONAL NIT: assign the function to a param above and then just do `err := filepath.Walk(roo, walkFn)` to make it easier to read.

File: tools/scripts/goimports/main.go [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+	if len(filesToProcess) > 0 {
+		cmd := exec.Command("goimports", append(defaultArgs, filesToProcess...)...)
+		if err := cmd.Run(); err != nil {
+			fmt.Printf("Failed running goimports: %v\n", err)
+		}
+	}
+}
+
+/*
+	containsEmptyImportScaffold checks if the go file at goSrcPath contains an
```
Comment (Olshansk) [inline tools/scripts/goimports/main.go:76]: Weird indent?

File: tools/scripts/goimports/main.go [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+		return false, err
+	}
+	defer file.Close()
+
+	scanner := bufio.NewScanner(file)
+	scanner.Split(importBlockSplit)
+
+	for scanner.Scan() {
+		block := scanner.Text()
+		if strings.Contains(block, "// this line is used by starport scaffolding # genesis/types/import") {
```
Comment (Olshansk) [inline tools/scripts/goimports/main.go:95]: Wdyt think?

```suggestion
		if strings.Contains(block, "// this line is used by starport scaffolding) {
```

File: tools/scripts/goimports/main.go [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+	if scanner.Err() != nil {
+		return false, scanner.Err()
+	}
+
+	return false, nil
+}
+
+// importBlockSplit is a split function intended to be used with bufio.Scanner
+// to extract the contents of a multi-line go import block.
+func importBlockSplit(data []byte, atEOF bool) (advance int, token []byte, err error) {
```
Comment (Olshansk) [inline tools/scripts/goimports/main.go:109]: 1. We don't use `atEOF` everywhere.
2. If we do end up using is, consider `isAtEOF` or `isEOF`

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: tools/scripts/goimports/main.go (hunk 3 above)
Comment (bryanchriswhite) [reply tools/scripts/goimports/main.go:14]: This is not something we're specifying, these are specified by, and passed to, the `goimports` binary.

File: tools/scripts/goimports/main.go (hunk 4 above)
Comment (bryanchriswhite) [reply tools/scripts/goimports/main.go:46]: Yessssss :+1: :rocket:

File: .github/workflows/go.yml [hunk 9]
```diff
@@ -35,6 +35,9 @@ jobs:
       - name: Generate mocks
         run: make go_mockgen
 
+      - name: Run golangci-lint
+        run: make go_lint
```
Comment (okdas) [inline .github/workflows/go.yml:39]: In CI, I would prefer us use an official [golangci-lint GitHub Action](https://github.com/marketplace/actions/run-golangci-lint). But I can hook it up in a subsequent PR if you would rather not deal with it. I was planning to add it along with our [custom linters from pocket repo](https://github.com/pokt-network/pocket/tree/main/build/linters). So thank you for taking the first step!

**Review by Olshansk — CHANGES_REQUESTED** [review]
@bryanchriswhite Please merge with main and will then re-review!

**Comment by bryanchriswhite** [comment]
> Please merge with main and will then re-review!

@Olshansk FWIW and future reference, had the `godoc_support`  branch not been deleted yet, the diff wouldn't have gone wonky in the time between merging #128 to main and merging main into this branch.

File: tools/scripts/goimports/filters/filters.go [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+)
+
+const igniteScaffoldComment = "// this line is used by starport scaffolding"
+
+var (
+	importStart = []byte("import (")
+	importEnd   = []byte(")")
+)
+
+// FilterFn  is a function that returns true if the given path matches the
```
Comment (Olshansk) [inline tools/scripts/goimports/filters/filters.go:20]:
```suggestion
// FilterFn is a function that returns true if the given path matches the
```

File: tools/scripts/goimports/filters/filters.go [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+
+// PathMatchesProtobufGatewayGo matches generated protobuf gateway go source files.
+func PathMatchesProtobufGatewayGo(path string) (bool, error) {
+	return strings.HasSuffix(path, ".pb.gw.go"), nil
+}
+
+// PathMatchesMockGo matches generated mock go source files.
+func PathMatchesMockGo(path string) (bool, error) {
+	return strings.HasSuffix(path, "_mock.go"), nil
+
```
Comment (Olshansk) [inline tools/scripts/goimports/filters/filters.go:42]:
```suggestion
```

**Review by Olshansk — APPROVED** [review]
Awesome work!

**Pull request merged**